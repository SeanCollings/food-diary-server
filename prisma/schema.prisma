// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

// https://www.prisma.io/nestjs
// https://docs.nestjs.com/recipes/prisma
// https://blog.logrocket.com/how-to-use-nestjs-prisma/

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model User {
  id        Int        @id @default(autoincrement())
  name      String
  email     String     @unique
  password  String
  avatar    String?
  createdAt DateTime   @default(now())
  lastLogin DateTime   @default(now())
  shareLink ShareLink?
  diaryDay  DiaryDay[]

  userPreferenceShowDayStreak       Boolean  @default(true)
  userPreferenceShowWeeklyExcercise Boolean?
  userPreferenceShowWeeklyWater     Boolean?

  statLastActivity DateTime?
  statDayStreak    Int?
}

model ShareLink {
  id       Int     @id @default(autoincrement())
  user     User    @relation(fields: [userId], references: [id])
  userId   Int     @unique
  link     String  @unique
  isShared Boolean
}

model DiaryDay {
  id                       String  @id @default(uuid())
  user                     User    @relation(fields: [userId], references: [id])
  userId                   Int
  date                     String
  mealBreakfast            Json?
  mealSnack1               Json?
  mealLunch                Json?
  mealSnack2               Json?
  mealDinner               Json?
  wellnessWater            Float?
  wellnessTeaCoffee        Float?
  wellnessAlcohol          Float?
  wellnessExcercise        String?
  wellnessExcerciseDetails String?

  hasMealBreakfast     Boolean?
  hasMealSnack1        Boolean?
  hasMealLunch         Boolean?
  hasMealSnack2        Boolean?
  hasMealDinner        Boolean?
  hasWellnessExcercise Boolean?

  @@unique([userId, date])
}
