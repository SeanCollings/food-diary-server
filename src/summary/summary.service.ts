import { formatSummaryData } from '@/utils/modules/summary-utils';
import { PrismaService } from '@/prisma.service';
import { Injectable } from '@nestjs/common';
import { getSortedDatesInRange } from '@/utils/date-utils';

@Injectable()
export class SummaryService {
  constructor(private prisma: PrismaService) {}

  async getUserSummary(userId: string, dateFrom: string, dateTo: string) {
    const datesInRange = getSortedDatesInRange(dateFrom, dateTo);

    const diaryDays = await this.prisma.diaryDay.findMany({
      where: {
        userId,
        date: { in: datesInRange },
      },
    });

    const summaryData = formatSummaryData(datesInRange, diaryDays);

    return {
      data: { ...summaryData },
      dates: datesInRange,
      totalDays: datesInRange.length,
    };
  }
}
