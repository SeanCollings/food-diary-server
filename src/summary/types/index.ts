import { MAX_SUMMARY_MONTH_RANGE } from '@/lib/constants/validation/validation.constants';
import { IsDateAfter } from '@/lib/validation/validators/is-date-after';
import { IsWithinDateRange } from '@/lib/validation/validators/is-within-date-range';
import { IsDateString, IsNotEmpty } from 'class-validator';

export interface RequestWithUser extends Request {
  user: { userId: number; email: string };
}

export class GetUserSummaryQuery {
  @IsDateString()
  @IsNotEmpty()
  dateFrom: string;

  @IsDateString()
  @IsNotEmpty()
  @IsDateAfter('dateFrom', { message: 'dateFrom must come before dateTo' })
  @IsWithinDateRange('dateFrom', {
    message: `dateFrom and dateTo max-range is ${MAX_SUMMARY_MONTH_RANGE} months`,
  })
  dateTo: string;
}
