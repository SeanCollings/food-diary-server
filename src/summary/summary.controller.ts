import { JwtAuthGuard } from '@/auth/jwt-auth.guard';
import { DEFAULT_ERROR_MSG } from '@/lib/validation/validation.constants';
import { Controller } from '@nestjs/common';
import {
  Get,
  InternalServerErrorException,
  UseGuards,
  Request,
  Query,
} from '@nestjs/common';
import { SummaryService } from './summary.service';
import { GetUserSummaryQuery, RequestWithUser } from './types';

@Controller('summary')
export class SummaryController {
  constructor(private summaryService: SummaryService) {}

  @Get()
  @UseGuards(JwtAuthGuard)
  async getUserSummary(
    @Request() req: RequestWithUser,
    @Query() query: GetUserSummaryQuery,
  ) {
    try {
      return await this.summaryService.getUserSummary(
        req.user.userId,
        query.dateFrom,
        query.dateTo,
      );
    } catch (err) {
      const errorMessage = err.message || DEFAULT_ERROR_MSG;
      console.error('[summary::_get_]:', errorMessage);
      throw new InternalServerErrorException(errorMessage);
    }
  }
}
