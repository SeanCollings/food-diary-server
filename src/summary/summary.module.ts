import { PrismaService } from '@/prisma.service';
import { Module } from '@nestjs/common';
import { SummaryController } from './summary.controller';
import { SummaryService } from './summary.service';

@Module({
  controllers: [SummaryController],
  providers: [SummaryService, PrismaService],
})
export class SummaryModule {}
