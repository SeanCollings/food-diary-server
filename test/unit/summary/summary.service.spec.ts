import { PrismaService } from '@/prisma.service';
import { SummaryService } from '@/summary/summary.service';
import { Test, TestingModule } from '@nestjs/testing';
import { DiaryDay, PrismaClient } from '@prisma/client';
import { DeepMockProxy, mockDeep } from 'jest-mock-extended';

describe('SummaryService', () => {
  const mockUserId = 'mock_user_id';
  let service: SummaryService;
  let prisma: DeepMockProxy<PrismaClient>;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [SummaryService, PrismaService],
    })
      .overrideProvider(PrismaService)
      .useValue(mockDeep<PrismaClient>())
      .compile();

    service = module.get(SummaryService);
    prisma = module.get(PrismaService);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('getUserSummary', () => {
    it('should return a user summary', async () => {
      prisma.diaryDay.findMany.mockResolvedValue([
        {
          id: 1,
          userId: mockUserId,
          date: '2023-04-13',
          mealBreakfast: [{ food: 'food_1' }] as any,
          mealSnack1: [{ food: 'food_2' }] as any,
          mealLunch: [{ food: 'food_3' }] as any,
          mealSnack2: [{ food: 'food_4' }] as any,
          mealDinner: undefined as any,
          wellnessWater: 0,
        },
      ] as unknown as DiaryDay[]);

      const result = await service.getUserSummary(
        mockUserId,
        '2023-04-12',
        '2023-04-13',
      );

      expect(result).toMatchSnapshot();
    });
  });
});
