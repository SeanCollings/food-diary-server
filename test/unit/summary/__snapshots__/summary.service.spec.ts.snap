// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SummaryService getUserSummary should return a user summary 1`] = `
{
  "data": {
    "2023-04-12T00:00:00.000Z": {
      "breakfast": [
        "food_1",
      ],
      "dinner": undefined,
      "lunch": [
        "food_3",
      ],
      "snack_1": [
        "food_2",
      ],
      "snack_2": [
        "food_4",
      ],
    },
  },
  "dates": [
    "2023-04-11T00:00:00.000Z",
    "2023-04-12T00:00:00.000Z",
  ],
  "totalDays": 2,
}
`;
